### PR TITLE
Type script

### PR DESCRIPTION
**Pull request by a99jafari**

### CODE REVIEW

File: src/components/AddTodo/index.tsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+interface IProps {
+    onAdd: (newTodo: ITodo) => void;
+}
+
+function AddTodo({ onAdd }: IProps): React.ReactElement {
+    const [todoTitle, setTodoTitle] = useState("");
+
+    const handleSubmit = (): void => {
```
Comment (sassanh) [inline src/components/AddTodo/index.tsx:17]: use `useCallback`

File: src/components/AddTodo/index.tsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        onAdd({
+            id: Math.floor(Math.random() * 10000),
+            title: todoTitle,
+            completed: false,
+        });
+        setTodoTitle("");
+    };
+
+    const handleChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
```
Comment (sassanh) [inline src/components/AddTodo/index.tsx:26]: use `useCallback` (we should use `useCallback` for all callbacks)

File: src/components/App/index.tsx [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import Todos from "../Todos";
+import AddTodo from "../AddTodo";
+import {
+    GlobalStyles,
+    FilterSection,
+    FilterButton,
+    MainSection,
+} from "./styles";
+
+interface ITodo {
```
Comment (sassanh) [inline src/components/App/index.tsx:11]: We don't use `I` prefix for interfaces

File: src/components/Todo/styles.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    }
+`;
+
+export const TodoTitle = styled.span`
+    font-family: "Lucida Sans", sans-serif;
+    font-size: 18px;
+    font-weight: bold;
+    vertical-align: middle;
+    color: ${(props: IProps) => (props.completed ? "#4fff38" : "#ffca38")};
+    text-decoration: ${(props) => (props.completed ? "line-through" : "none")};
```
Comment (sassanh) [inline src/components/Todo/styles.ts:47]:
```suggestion
    text-decoration: ${({completed}) => (completed ? "line-through" : "none")};
```

**Review by sassanh — CHANGES_REQUESTED** [review]

File: src/components/AddTodo/index.tsx (hunk 2 above)
Comment (a99jafari) [reply src/components/AddTodo/index.tsx:26]: When I use `useCallback` here, it seems there is have no necessary dependency. What is our dependency here?

Comment (sassanh) [reply src/components/AddTodo/index.tsx]: We have no dependency which means we can use the same callback from the moment this component is mounted until it has been unmounted without ever changing it.

File: src/components/App/index.tsx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+                return todos;
+        }
+    }, [filterName, todos]);
+
+    return (
+        <MainSection>
+            {GlobalStyles}
+            <AddTodo onAdd={handleAdd} />
+            <FilterSection>
+                <FilterButton onClick={() => setFilterName("ALL")}>
```
Comment (sassanh) [inline src/components/App/index.tsx:101]: As we already talked **"no callback should be written inline or without useCallback"** please fix these and make sure no callback in your code is being used without `useCallback`

File: src/components/App/index.tsx [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    const handleFilter = useCallback((filterTitle) => {
+        setFilterName(filterTitle);
+    }, []);
+
+    return (
+        <MainSection>
+            {GlobalStyles}
+            <AddTodo onAdd={handleAdd} />
+            <FilterSection>
+                <FilterButton onClick={handleFilter.bind(null, "All")}>
```
Comment (sassanh) [inline src/components/App/index.tsx:105]: Using `.bind` will create a new instance of the function on each render, so even though `handleFilter` will remain the same in each render, `handleFilter.bind...` will be changed in each render.

**Review by sassanh — APPROVED** [review]

**Pull request merged**